### ISSUE TITLE
Proposal: Improve C++ tooling.

### ISSUE THREAD

**Issue by pjmlp** [microsoft/microsoft-ui-xaml#4082]
Currently C++/WinRT developer experience still feels backward versus what C++/CX offers, the pain points being:

- IDL files are required
- no Visual Studio tooling support for IDL files, which turns the experience akin to using plain old Notepad
- added burden to manually compile those IDL files and then merge their contents into existing code
- although the documentation has been greatly improved, it still is quite overwhelming what one needs to understand to properly write full blown applications without landing into C++/CX samples or figuring out how to actually write some stuff in C++/WinRT.

Team should spend some time looking at cross platform C++ frameworks that offer first class developer experience if the goal is for us still care about Windows specific GUI frameworks.

Here are some examples that excel in C++ developer tools for GUI development.

https://www.noesisengine.com/index.php (with support for XAML even)

https://www.qt.io/product/development-tools

https://www.embarcadero.com/products/cbuilder

**Comment by StephenLPeters** [comment]
@kennykerr FYI

**Comment by kennykerr** [comment]
Agreed - @Scottj1s and I worked on a prototype for getting rid of IDL a while back. [It showed a lot of promise](https://www.youtube.com/watch?v=X41j_gzSwOY).

**Comment by pjmlp** [comment]
I am aware of that session, basically the _"we should put up with C++/WinRT tooling and wait for the day C++ gets reflection"_ answer, or just ignore WinUI and go cross platform with one of the previously mentioned C++ frameworks from the competition.

**Comment by kennykerr** [comment]
> wait for the day C++ gets reflection

Actually that demo illustrated how it could be done with today's C++, but it still requires some work to complete that experience.

**Comment by pjmlp** [comment]
Good news then, anything that can be shared besides just hoping C++/WinRT will ever match the cross-platform competition?

**Comment by pjmlp** [comment]
One year and half have gone by, and the C++/WinRT tooling in Visual Studio hasn't changed a bit, is there any hope to actually be able to enjoy a  **_modern_** tooling experience for WinUI apps with C++?

As it is, C++ Builder and QtCreator seem to be the only way forward.

**Comment by Gavin-Williams** [comment]
Has anything changed from 3 years ago? I tried using C++/WinRT yesterday and couldn't get even the simplest usage working. Code wouldn't compile but there were zero errors - just one snapshot of the issues in had in the day. It seems like C++/WinRT is too fragile and completely hostile to beginners and casual users.

**Comment by YexuanXiao** [comment]
Except for C++/WinRT, no other UI library provides first-class support for coroutines. I'm curious why you're eager to use C++26 yet overlook this advantage of C++/WinRT. Using any library that doesn't support coroutines is a regression compared to C++/WinRT

**Comment by pjmlp** [comment]
If we ignore .NET that is, which was used for Microsoft's original proposal to C++20, based on previous work with C++/CX.

C++26 support on VC++? Good luck with that.

https://developercommunity.visualstudio.com/t/Implement-C26-Standard-features-in-MSV/10777423

**Comment by YexuanXiao** [comment]
Which C++26 feature do you want? Reflection? I don't think it can completely replace IDL. I've been following most of the C++26 features. They're nice, but not enough to redefine what C++ projections should be.

Clang isn't always good, although it often gets new cool features faster because it's open source. For example, I found that Clang 20 x86 generates incorrect code for lambdas, which caused me to disable it in a test for my C++/WinRT PR.

**Comment by YexuanXiao** [comment]
I care more about two things. First, raise the C++ standard used by C++/WinRT and allow some breaking changes. Incremental improvements are better than big-bang changes. Second, make XAML and WAS accept external contributions. XAML is already close to that, while WAS is still far off. Especially WAS, because its code is worse.
